File: src/test/java/no/nav/sosialhjelp/soknad/business/db/repositories/soknadunderarbeid/BatchSoknadUnderArbeidRepositoryJdbcTest.java
package no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid;

import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad;
import no.nav.sosialhjelp.soknad.business.db.RepositoryTestSupport;
import no.nav.sosialhjelp.soknad.business.db.repositories.opplastetvedlegg.BatchOpplastetVedleggRepository;
import no.nav.sosialhjelp.soknad.business.db.repositories.opplastetvedlegg.OpplastetVedleggRepository;
import no.nav.sosialhjelp.soknad.config.DbTestConfig;
import no.nav.sosialhjelp.soknad.domain.OpplastetVedlegg;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.domain.VedleggType;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import javax.inject.Inject;
import java.time.LocalDateTime;

import static no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeidStatus.UNDER_ARBEID;
import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = {DbTestConfig.class})
@ActiveProfiles("test")
class BatchSoknadUnderArbeidRepositoryJdbcTest {

    private static final String EIER = "12345678901";
    private static final String BEHANDLINGSID = "1100020";
    private static final String TILKNYTTET_BEHANDLINGSID = "4567";
    private static final JsonInternalSoknad JSON_INTERNAL_SOKNAD = new JsonInternalSoknad();

    @Inject
    private RepositoryTestSupport soknadRepositoryTestSupport;

    @Inject
    private SoknadUnderArbeidRepository soknadUnderArbeidRepository;

    @Inject
    private OpplastetVedleggRepository opplastetVedleggRepository;

    @Inject
    private BatchSoknadUnderArbeidRepository batchSoknadUnderArbeidRepository;

    @Inject
    private BatchOpplastetVedleggRepository batchOpplastetVedleggRepository;

    @AfterEach
    public void tearDown() {
        soknadRepositoryTestSupport.getJdbcTemplate().update("delete from SOKNAD_UNDER_ARBEID");
        soknadRepositoryTestSupport.getJdbcTemplate().update("delete from OPPLASTET_VEDLEGG");
    }

    @Test
    void hentSoknaderForBatchSkalFinneGamleSoknader() {
        var skalIkkeSlettes = lagSoknadUnderArbeid(BEHANDLINGSID, 13);
        var skalIkkeSlettesId = soknadUnderArbeidRepository.opprettSoknad(skalIkkeSlettes, EIER);

        var skalSlettes = lagSoknadUnderArbeid("annen_behandlingsid", 14);
        var skalSlettesId = soknadUnderArbeidRepository.opprettSoknad(skalSlettes, EIER);

        var soknader = batchSoknadUnderArbeidRepository.hentGamleSoknadUnderArbeidForBatch();

        assertThat(soknader).hasSize(1);
        assertThat(soknader.get(0)).isEqualTo(skalSlettesId);
    }

    @Test
    void slettSoknadGittSoknadUnderArbeidIdSkalSletteSoknad() {
        var soknadUnderArbeid = lagSoknadUnderArbeid(BEHANDLINGSID, 15);
        var soknadUnderArbeidId = soknadUnderArbeidRepository.opprettSoknad(soknadUnderArbeid, EIER);
        soknadUnderArbeid.setSoknadId(soknadUnderArbeidId);
        var opplastetVedleggUuid = opplastetVedleggRepository.opprettVedlegg(lagOpplastetVedlegg(soknadUnderArbeidId), EIER);

        batchSoknadUnderArbeidRepository.slettSoknad(soknadUnderArbeid.getSoknadId());

        assertThat(soknadUnderArbeidRepository.hentSoknad(soknadUnderArbeidId, EIER)).isEmpty();
        assertThat(opplastetVedleggRepository.hentVedlegg(opplastetVedleggUuid, EIER)).isEmpty();
    }

    private SoknadUnderArbeid lagSoknadUnderArbeid(String behandlingsId, int antallDagerSiden) {
        return new SoknadUnderArbeid().withVersjon(1L)
                .withBehandlingsId(behandlingsId)
                .withTilknyttetBehandlingsId(TILKNYTTET_BEHANDLINGSID)
                .withEier(EIER)
                .withJsonInternalSoknad(JSON_INTERNAL_SOKNAD)
                .withStatus(UNDER_ARBEID)
                .withOpprettetDato(LocalDateTime.now().minusDays(antallDagerSiden).minusMinutes(5))
                .withSistEndretDato(LocalDateTime.now().minusDays(antallDagerSiden).minusMinutes(5));
    }

    private OpplastetVedlegg lagOpplastetVedlegg(Long soknadId) {
        return new OpplastetVedlegg()
                .withEier(EIER)
                .withVedleggType(new VedleggType("bostotte|annetboutgift"))
                .withData(new byte[]{1, 2, 3})
                .withSoknadId(soknadId)
                .withFilnavn("dokumentasjon.pdf")
                .withSha512("aaa");
    }

}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/repository/soknadunderarbeid/BatchSoknadUnderArbeidRepositoryJdbcTest.kt
package no.nav.sosialhjelp.soknad.repository.soknadunderarbeid

import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.BatchOpplastetVedleggRepository
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.BatchOpplastetVedleggRepositoryJdbc
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.OpplastetVedlegg
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.OpplastetVedleggRepository
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.OpplastetVedleggRepositoryJdbc
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.OpplastetVedleggType
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.BatchSoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.BatchSoknadUnderArbeidRepositoryJdbc
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepositoryJdbc
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidStatus
import no.nav.sosialhjelp.soknad.repository.RepositoryTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.transaction.support.TransactionTemplate
import java.time.LocalDateTime

internal class BatchSoknadUnderArbeidRepositoryJdbcTest : RepositoryTest() {
    
    private var soknadUnderArbeidRepository: SoknadUnderArbeidRepository? = null
    private var opplastetVedleggRepository: OpplastetVedleggRepository? = null
    private var batchSoknadUnderArbeidRepository: BatchSoknadUnderArbeidRepository? = null
    private var batchOpplastetVedleggRepository: BatchOpplastetVedleggRepository? = null

    @Autowired
    private lateinit var transactionTemplate: TransactionTemplate

    @BeforeEach
    fun setup() {
        if (opplastetVedleggRepository == null) {
            opplastetVedleggRepository = OpplastetVedleggRepositoryJdbc(jdbcTemplate)
        }
        if (soknadUnderArbeidRepository == null) {
            soknadUnderArbeidRepository = SoknadUnderArbeidRepositoryJdbc(
                jdbcTemplate, transactionTemplate, opplastetVedleggRepository!!
            )
        }
        if (batchOpplastetVedleggRepository == null) {
            batchOpplastetVedleggRepository = BatchOpplastetVedleggRepositoryJdbc(jdbcTemplate)
        }
        if (batchSoknadUnderArbeidRepository == null) {
            batchSoknadUnderArbeidRepository = BatchSoknadUnderArbeidRepositoryJdbc(
                jdbcTemplate,transactionTemplate, batchOpplastetVedleggRepository!!)
        }
    }

    @Test
    fun hentSoknaderForBatchSkalFinneGamleSoknader() {
        val skalIkkeSlettes = lagSoknadUnderArbeid(BEHANDLINGSID, 13)
        val skalIkkeSlettesId = soknadUnderArbeidRepository!!.opprettSoknad(skalIkkeSlettes, EIER)
        val skalSlettes = lagSoknadUnderArbeid("annen_behandlingsid", 14)
        val skalSlettesId = soknadUnderArbeidRepository!!.opprettSoknad(skalSlettes, EIER)
        val soknader = batchSoknadUnderArbeidRepository!!.hentGamleSoknadUnderArbeidForBatch()
        assertThat(soknader).hasSize(1)
        assertThat(soknader[0]).isEqualTo(skalSlettesId).isNotEqualTo(skalIkkeSlettesId)
    }

    @Test
    fun slettSoknadGittSoknadUnderArbeidIdSkalSletteSoknad() {
        val soknadUnderArbeid = lagSoknadUnderArbeid(BEHANDLINGSID, 15)
        val soknadUnderArbeidId = soknadUnderArbeidRepository!!.opprettSoknad(soknadUnderArbeid, EIER)
        soknadUnderArbeid.soknadId = soknadUnderArbeidId!!
        val opplastetVedleggUuid = opplastetVedleggRepository!!.opprettVedlegg(lagOpplastetVedlegg(soknadUnderArbeidId), EIER)
        batchSoknadUnderArbeidRepository!!.slettSoknad(soknadUnderArbeid.soknadId)
        assertThat(soknadUnderArbeidRepository!!.hentSoknad(soknadUnderArbeidId, EIER)).isNull()
        assertThat(opplastetVedleggRepository!!.hentVedlegg(opplastetVedleggUuid, EIER)).isNull()
    }

    private fun lagSoknadUnderArbeid(behandlingsId: String, antallDagerSiden: Int): SoknadUnderArbeid {
        return SoknadUnderArbeid(
            versjon = 1L,
            behandlingsId = behandlingsId,
            tilknyttetBehandlingsId = TILKNYTTET_BEHANDLINGSID,
            eier = EIER,
            jsonInternalSoknad = JSON_INTERNAL_SOKNAD,
            status = SoknadUnderArbeidStatus.UNDER_ARBEID,
            opprettetDato = LocalDateTime.now().minusDays(antallDagerSiden.toLong()).minusMinutes(5),
            sistEndretDato = LocalDateTime.now().minusDays(antallDagerSiden.toLong()).minusMinutes(5)
        )
    }

    private fun lagOpplastetVedlegg(soknadId: Long): OpplastetVedlegg {
        return OpplastetVedlegg(
            eier = EIER,
            vedleggType = OpplastetVedleggType("bostotte|annetboutgift"),
            data = byteArrayOf(1, 2, 3),
            soknadId = soknadId,
            filnavn = "dokumentasjon.pdf",
            sha512 = "aaa"
        )
    }

    companion object {
        private const val EIER = "12345678901"
        private const val BEHANDLINGSID = "1100020"
        private const val TILKNYTTET_BEHANDLINGSID = "4567"
        private val JSON_INTERNAL_SOKNAD = JsonInternalSoknad()
    }
}
